component/common/net/server.go

```diff
@@ -1,97 +1,353 @@
-		metricsNamespace: metricsNamespace,
-	}
-
+	// Apply some defaults if nothing provided
 	if config == nil {
 		config = &ServerConfig{}
+		// Set the config to the new combined config.
+		// Avoid logging entire received request on failures
+		config.ExcludeRequestInLog = true
+		// Configure dedicated metrics registerer
+		config.Registerer = reg
+		// To prevent metric collisions because all metrics are going to be registered in the global Prometheus registry.
+		config.MetricsNamespace = metricsNamespace
+		// We don't want the /debug and /metrics endpoints running, since this is not the main Flow HTTP server.
+		// We want this target to expose the least surface area possible, hence disabling WeaveWorks HTTP server metrics
+		// and debugging functionality.
+		config.RegisterInstrumentation = false
+		// Add logger to weaveworks
+		config.Log = logging.GoKit(logger)
+
 	}
 
-	// convert from River into the weaveworks config
-	serverCfg := config.Convert()
-	// Set the config to the new combined config.
-	// Avoid logging entire received request on failures
-	serverCfg.ExcludeRequestInLog = true
-	// Configure dedicated metrics registerer
-	serverCfg.Registerer = reg
-	// Persist crafter config in server
-	ts.config = &serverCfg
-	// To prevent metric collisions because all metrics are going to be registered in the global Prometheus registry.
-	ts.config.MetricsNamespace = ts.metricsNamespace
-	// We don't want the /debug and /metrics endpoints running, since this is not the main Flow HTTP server.
-	// We want this target to expose the least surface area possible, hence disabling WeaveWorks HTTP server metrics
-	// and debugging functionality.
-	ts.config.RegisterInstrumentation = false
-	// Add logger to weaveworks
-	ts.config.Log = logging.GoKit(ts.logger)
-
-	return ts, nil
+	// Apply some defaults if nothing provided
+	if config.HTTP == nil {
+		config.HTTP = &HTTPConfig{ListenPort: DefaultHTTPPort}
+	}
+
+	return New(*config)
 }
 
-// MountAndRun mounts the handlers and starting the server.
-func (ts *TargetServer) MountAndRun(mountRoute func(router *mux.Router)) error {
-	level.Info(ts.logger).Log("msg", "starting server")
-	srv, err := weaveworks.New(*ts.config)
+// New makes a new Server
+func New(cfg ServerConfig) (*Server, error) {
+	// If user doesn't supply a logging implementation, by default instantiate
+	// logrus.
+	logger := cfg.Log
+	if logger == nil {
+		logger = logging.NewLogrus(cfg.LogLevel)
+	}
+
+	// If user doesn't supply a registerer/gatherer, use Prometheus' by default.
+	reg := cfg.Registerer
+	if reg == nil {
+		reg = prometheus.DefaultRegisterer
+	}
+	gatherer := cfg.Gatherer
+	if gatherer == nil {
+		gatherer = prometheus.DefaultGatherer
+	}
+
+	tcpConnections := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: cfg.MetricsNamespace,
+		Name:      "tcp_connections",
+		Help:      "Current number of accepted TCP connections.",
+	}, []string{"protocol"})
+	reg.MustRegister(tcpConnections)
```

Comment:
We will now be able to use `Register` and reuse the existing metric if it's already registered. Avoiding the need for workaround here https://github.com/grafana/agent/pull/3711